drivers/led/lp5521.c

```diff
@@ -0,0 +1,910 @@
+			       uint8_t step_count,
+			       enum lp5521_engine_fade_dirs fade_dir)
+{
+	struct lp5521_data *data = dev->data;
+	struct led_data *dev_data = &data->dev_data;
+	uint8_t prescale, step_time;
+
+	if ((time_per_step < dev_data->min_period) ||
+			(time_per_step > dev_data->max_period)) {
+		return -EINVAL;
+	}
+
+	lp5521_ms_to_prescale_and_step(dev_data, time_per_step,
+			&prescale, &step_time);
+
+	return lp5521_program_command(dev, led, command_index,
+			LP5521_PROG_COMMAND_RAMP_TIME(prescale, step_time),
+			LP5521_PROG_COMMAND_STEP_COUNT(fade_dir, step_count));
+}
+
+/*
+ * @brief Program a command to do nothing for the given time.
+ *
+ * @param dev           LP5521 device.
+ * @param led           target led.
+ * @param command_index Index of the command in the program sequence.
+ * @param time          Time to do nothing in milliseconds.
+ *
+ * @retval 0       On success.
+ * @retval -EINVAL If the passed arguments are invalid or out of range.
+ * @retval -EIO    If the underlying I2C call fails.
+ */
+static inline int lp5521_program_wait(const struct device *dev,
+						enum lp5521_led_channels led,
+						uint8_t command_index,
+						uint32_t time)
+{
+	/*
+	 * A wait command is a ramp with the step_count set to 0. The fading
+	 * direction does not matter in this case.
+	 */
+	return lp5521_program_ramp(dev, led, command_index,
+			time, 0, LP5521_FADE_UP);
+}
+
+/*
+ * @brief Program a command to go back to the beginning of the program.
+ *
+ * Can be used at the end of a program to loop it infinitely.
+ *
+ * @param dev           LP5521 device.
+ * @param led           target led
+ * @param command_index Index of the command in the program sequence.
+ *
+ * @retval 0       On success.
+ * @retval -EINVAL If the given command index is out of range or an invalid
+ *		   engine is passed.
+ * @retval -EIO    If the underlying I2C call fails.
+ */
+static inline int lp5521_program_go_to_start(const struct device *dev,
+							enum lp5521_led_channels led,
+							uint8_t command_index)
+{
+	return lp5521_program_command(dev, led, command_index, 0x00, 0x00);
+}
+
+/*
+ * @brief Change the brightness of a running blink program.
+ *
+ * We know that the current program executes a blinking pattern
+ * consisting of following commands:
+ *
+ * - set_brightness high
+ * - wait on_delay
+ * - set_brightness low
+ * - wait off_delay
+ * - return to start
+ *
+ * In order to change the brightness during blinking, we overwrite only
+ * the first command and start execution again.
+ *
+ * @param dev           LP5521 device.
+ * @param led           target led.
+ * @param brightness_on New brightness value.
+ *
+ * @retval 0       On Success.
+ * @retval -EINVAL If the engine ID or brightness is out of range.
+ * @retval -EIO    If the underlying I2C call fails.
+ */
+static int lp5521_update_blinking_brightness(const struct device *dev,
+							enum lp5521_led_channels led,
+							uint8_t brightness_on)
```

Comment:
Let's fix formatting on Function declarations/definitions throughout the whole PR (just capturing it once)